### DISCUSSION TITLE
measure power

### DISCUSSION THREAD

**Discussion by btwbtw01** [Xilinx/finn#510] [Q&A] [answered]
Hi,
Can I measure the power or energy only DRAM and PL side? Our application is to compress activation to reduce DRAM access power. Does Xilinx have the tool to support? Maybe Xilinx Power Estimator?

Has anybody ever used the tool to measure power and energy that comes from Xilinx?

Thanks!!

**Comment by fpjentzsch** [comment, marked as answer]
Hi, which board are you targeting? For some development boards (e.g. ZCU102) you can measure power consumption of various rails by reading out some board components via the I2C/PMBUS:

1. The power controller chips: Works even with an external device ("MAXPOWERTOOL") you connect to the PMBUS header, but is horribly inaccurate for the ZCU102 due to a board design issue.
2. The power monitor chips: Fairly accurate, can be set-up and polled via PS (software), PL, or the system controller if your board has one.

I can't help you with power estimation during design time, but I'd also like to know how accurate this can get.

**Reply by btwbtw01** [reply]
I'm using ZCU104 because it can support PYNQ OS. I would like to know how you measure power in the FINN or FINN-R paper. How can I use power monitor chips to measure power. 
By the way, do you know power advantage tool? Have you ever used it before?

**Reply by btwbtw01** [reply]
Hi, @fpjentzsch 
Can I read the power consumption of each array like FIFO, compute array of each layer when we synthesize and generate IP? I remember that we can read power consumption of each IP in vivado?

**Reply by fpjentzsch** [reply]
I have never used the Power Advantage Tool directly, as it is more of a demo. Internally, it probes the power monitor chips (INA226) via the MSP430 microcontroller (system controller). I have measured ZCU102 power in a similar fashion before, but via the [system controller GUI (SCUI, aka. "Board GUI")](https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/1544716298/Zynq+Ultrascale+MPSOC+BIST+and+SCUI+Guide).
At a glance, the ZCU104's power control/monitoring circuitry is setup differently, but I'm not familiar with it.

Regarding your 2nd question: Yes, in the Vivado-generated power report, you can see power consumption per IP (or even HLS function) unter the "Utilization Details -> Hierarchical" tab. However, I've found Vivado power reports to be very inaccurate (off by 2x-3x), at least if you don't provide simulation-based switching activity information.

Btw, it is also possible to build PYNQ OS images for other boards, including the ZCU102.

**Reply by btwbtw01** [reply]
Hi, @fpjentzsch 
I'm using system controller GUI before, but it can only measure one time. It can't measure power constatly. However I didn't use power monitor chips (INA226). I'm following the manual which I download from Xilinx
[xtp433-zcu102-system-controller-c-2019-1.pdf](https://github.com/Xilinx/finn/files/8129602/xtp433-zcu102-system-controller-c-2019-1.pdf)
ZCU104 also has system controller GUI https://www.xilinx.com/support/documentation/boards_and_kits/zcu104/2019_1/xtp502-zcu104-system-controller-c-2019-1.pdf
How to build PYNQ OS images for ZCU102? I try to build PYNQ OS images for ZC706 but failed.

**Reply by btwbtw01** [reply]
Hi, @fpjentzsch 
Talk about Vivado power reports, simulation-based switching activity information means that I can feed input data to get switching activity information, so that the report will be accurate.

**Reply by fpjentzsch** [reply]
I believe the official guide to build PYNQ images (https://pynq.readthedocs.io/en/latest/pynq_sd_card.html) worked for me in the past.

Yes, you will need to simulate the design in Vivado with realistic input data. Then you can save switching activity information to a file. However, you can't do that right out-of-the-box with FINN because we do RTL simulation outside of Vivado with Verilator.

**Reply by 218w1d7706** [reply]
hi, @fpjentzsch 
is it good to use SCUI for power monitoring or i have to use use power advantage tool?
thank you.

**Reply by Ba1tu3han** [reply]
@fpjentzsch 

Hello Felix,

Similarly I want to measure total power consumption of my HW design which is a CNV-like model in Vivado. I completed succesfully all steps of FINN and implemented in PYNQ-Z2 (does not have PMBUS). 

The power analysis does not make sense and it has low confidence level. 

What do you recommend to measure total power to compare different bit-width variation of the same NN model?

Bests,

**Reply by fpjentzsch** [reply]
Hi, you could measure the total board power going into the PYNQ-Z2 and see if you can detect meaningful differences between the models while they are running continuously. I prefer to use the onboard current measurement infrastructure (via I2C/PMBUS), but small boards like the Z2 often lack this feature.

I have also heard of people soldering their own shunt resistors onto PYNQ boards to measure the main power rail with an external meter...

**Reply by Ba1tu3han** [reply]
Hi  Felix,

Thank you for your recommendation.

I am able to calculate the theoretical power consumption on Vivado by using the Vivado Project file which is generated by FINN. It gives an general idea.

I will consider using a shunt resistor.

Bests,
